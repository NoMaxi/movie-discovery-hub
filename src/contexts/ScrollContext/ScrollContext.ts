import { createContext } from "react";

export interface ScrollContextType {
    targetMovieId: number | null;
    setTargetMovieId: (id: number | null) => void;
    triggerScroll: boolean;
    setTriggerScroll: (trigger: boolean) => void;
}

export const ScrollContext = createContext<ScrollContextType | undefined>(undefined);
