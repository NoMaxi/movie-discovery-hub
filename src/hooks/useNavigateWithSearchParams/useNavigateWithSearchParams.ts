import { useCallback } from "react";
import { useNavigate, useSearchParams } from "react-router-dom";

export const useNavigateWithSearchParams = () => {
    const navigate = useNavigate();
    const [searchParams] = useSearchParams();

    return useCallback(
        (path: string) => {
            const searchParamsString = searchParams.toString();
            navigate(`${path}${searchParamsString ? `?${searchParamsString}` : ""}`);
        },
        [navigate, searchParams],
    );
};
