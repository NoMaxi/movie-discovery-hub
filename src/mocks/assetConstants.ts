// Mock version of `src/constants/assetConstants.ts` for tests (mock images imports)
export const DEFAULT_IMAGE_URL = "test-image-stub";
