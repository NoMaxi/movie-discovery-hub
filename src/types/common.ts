export type Genre = "All" | "Documentary" | "Comedy" | "Horror" | "Crime";

export interface Movie {
    id: string;
    imageUrl: string;
    title: string;
    releaseYear: number;
    genres: string[];
}
