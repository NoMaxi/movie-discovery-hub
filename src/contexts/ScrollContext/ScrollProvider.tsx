import { useState, ReactNode } from "react";
import { ScrollContext } from "./ScrollContext";

export const ScrollProvider = ({ children }: { children: ReactNode }) => {
    const [targetMovieId, setTargetMovieId] = useState<number | null>(null);
    const [triggerScroll, setTriggerScroll] = useState<boolean>(false);

    return (
        <ScrollContext.Provider value={{ targetMovieId, setTargetMovieId, triggerScroll, setTriggerScroll }}>
            {children}
        </ScrollContext.Provider>
    );
};
