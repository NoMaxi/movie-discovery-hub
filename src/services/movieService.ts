import axios, { AxiosResponse } from "axios";
import { GenreFilter, Movie, SortOption } from "@/types/common";
import { DEFAULT_MOVIES_PER_PAGE } from "@/constants/constants";
import { mapAPIMovieDetailsToMovie, mapSortOptionToSortBy, mapSortOptionToSortOrder } from "@/utils/movieMapping";

export interface CreateMoviePayload {
    genres: string[];
    overview: string;
    poster_path: string;
    runtime: number;
    title: string;
    budget?: number;
    release_date?: string;
    revenue?: number;
    tagline?: string;
    vote_average?: number;
    vote_count?: number;
}

export interface APIMovieDetails extends Omit<CreateMoviePayload, "runtime"> {
    id: number;
    runtime: number | null;
}

export interface MoviesResponse {
    data: APIMovieDetails[];
    totalAmount: number;
    offset: number;
    limit: number;
}

export type SearchBy = "title" | "genres";
export type SortBy = "title" | "release_date";
export type SortOrder = "asc" | "desc";

export interface MoviesRequestParams {
    searchBy?: SearchBy;
    sortBy?: SortBy;
    sortOrder?: SortOrder;
    search?: string;
    limit?: number;
    offset?: number;
    filter?: GenreFilter;
}

export interface InfiniteMovieListResult {
    movies: Movie[];
    totalAmount: number;
    offset: number;
    limit: number;
    nextOffset?: number;
}

export const API_URL = "http://localhost:4000/movies";

export const movieService = {
    async getMovies(
        searchQuery: string,
        sortCriterion: SortOption,
        activeGenre: GenreFilter,
        signal: AbortSignal | undefined,
        offset: number = 0,
    ): Promise<InfiniteMovieListResult> {
        const params: MoviesRequestParams = {
            sortBy: mapSortOptionToSortBy(sortCriterion),
            sortOrder: mapSortOptionToSortOrder(sortCriterion),
            searchBy: "title",
            limit: DEFAULT_MOVIES_PER_PAGE,
            offset,
        };

        if (searchQuery) {
            params.search = searchQuery;
        }

        if (activeGenre !== "All") {
            params.filter = activeGenre;
        }

        const { data: moviesResponse } = await axios.get<MoviesResponse>(API_URL, { params, signal });

        const { offset: responseOffset, totalAmount, limit, data } = moviesResponse;
        const movies = data.map(mapAPIMovieDetailsToMovie);
        const currentTotalFetched = offset + movies.length;
        const nextOffset = currentTotalFetched < totalAmount ? currentTotalFetched : undefined;

        return {
            movies,
            totalAmount,
            limit,
            offset: responseOffset,
            nextOffset,
        };
    },

    async getAPIMovieDetailsById(id: number): Promise<APIMovieDetails> {
        const response = await axios.get<APIMovieDetails>(`${API_URL}/${id}`);
        return response.data;
    },

    async getMovieById(id: number): Promise<Movie> {
        const movieDetails = await this.getAPIMovieDetailsById(id);
        return mapAPIMovieDetailsToMovie(movieDetails);
    },

    async createMovie(movieData: CreateMoviePayload): Promise<Movie> {
        const { data } = await axios.post<APIMovieDetails, AxiosResponse<APIMovieDetails>, CreateMoviePayload>(
            API_URL,
            movieData,
        );
        return mapAPIMovieDetailsToMovie(data);
    },

    async updateMovie(movieData: APIMovieDetails): Promise<Movie> {
        const { data } = await axios.put<APIMovieDetails, AxiosResponse<APIMovieDetails>, APIMovieDetails>(
            API_URL,
            movieData,
        );
        return mapAPIMovieDetailsToMovie(data);
    },

    async deleteMovie(id: number): Promise<void> {
        await axios.delete(`${API_URL}/${id}`);
    },
};
