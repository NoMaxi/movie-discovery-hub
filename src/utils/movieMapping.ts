import { InitialMovieInfo, Movie, SortOption } from "@/types/common";
import { DEFAULT_DESCRIPTION, DEFAULT_TITLE } from "@/constants/constants";
import { DEFAULT_IMAGE_URL } from "@/constants/assetConstants";
import { getYearFromDate } from "@/utils/formatting";
import { APIMovieDetails, SortBy, SortOrder } from "@/services/movieService";

export const mapAPIMovieDetailsToMovie = ({
    id,
    genres,
    overview,
    poster_path,
    release_date,
    runtime,
    title,
    vote_average,
}: APIMovieDetails): Movie => ({
    id,
    description: overview || DEFAULT_DESCRIPTION,
    duration: runtime ?? 0,
    genres,
    imageUrl: poster_path || DEFAULT_IMAGE_URL,
    rating: vote_average ?? 0,
    releaseYear: getYearFromDate(release_date),
    title: title || DEFAULT_TITLE,
});

export const mapAPIMovieDetailsToInitialMovieInfo = (apiMovieDetails: APIMovieDetails): InitialMovieInfo => {
    const movie = mapAPIMovieDetailsToMovie(apiMovieDetails);
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const { releaseYear, ...movieWithoutReleaseYear } = movie;

    return {
        ...movieWithoutReleaseYear,
        releaseDate: apiMovieDetails.release_date ?? "",
    };
};

const SORT_OPTION_TO_SORT_BY: Record<SortOption, SortBy> = {
    "Release Date": "release_date",
    Title: "title",
};

const SORT_OPTION_TO_SORT_ORDER: Record<SortOption, SortOrder> = {
    "Release Date": "desc",
    Title: "asc",
};

export const mapSortOptionToSortBy = (sortOption: SortOption): SortBy => {
    return SORT_OPTION_TO_SORT_BY[sortOption] || "release_date";
};

export const mapSortOptionToSortOrder = (sortOption: SortOption): SortOrder => {
    return SORT_OPTION_TO_SORT_ORDER[sortOption] || "desc";
};
