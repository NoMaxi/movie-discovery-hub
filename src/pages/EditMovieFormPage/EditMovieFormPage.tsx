import { useEffect, useState } from "react";
import { useParams } from "react-router-dom";
import { InitialMovieInfo, MovieFormData } from "@/types/common";
import { mapAPIMovieDetailsToInitialMovieInfo } from "@/utils/movieMapping";
import { movieService } from "@/services/movieService";
import { useScrollContext } from "@/contexts/ScrollContext/useScrollContext";
import { useNavigateWithSearchParams } from "@/hooks/useNavigateWithSearchParams/useNavigateWithSearchParams";
import { Dialog } from "@/components/common/Dialog/Dialog";
import { Loader } from "@/components/common/Loader/Loader";
import { ErrorMessage } from "@/components/common/ErrorMessage/ErrorMessage";
import { MovieForm } from "@/components/MovieForm/MovieForm";

export const EditMovieFormPage = () => {
    const navigateWithSearchParams = useNavigateWithSearchParams();
    const { movieId } = useParams<{ movieId: string }>();
    const { setTriggerScroll } = useScrollContext();

    const [isFetchingDetails, setIsFetchingDetails] = useState(true);
    const [isSubmitting, setIsSubmitting] = useState(false);
    const [errorMessage, setErrorMessage] = useState<string | null>(null);
    const [movieInfo, setMovieInfo] = useState<InitialMovieInfo | null>(null);

    useEffect(() => {
        const fetchMovieDetails = async () => {
            if (!movieId) {
                setIsFetchingDetails(false);
                setErrorMessage("Movie ID not found.");
                return;
            }

            setIsFetchingDetails(true);
            setErrorMessage(null);

            try {
                const movie = await movieService.getAPIMovieDetailsById(Number(movieId));
                setMovieInfo(mapAPIMovieDetailsToInitialMovieInfo(movie));
            } catch (err) {
                console.error("Error fetching movie details during movie edit:", err);
                setErrorMessage("Failed to load movie details. Please try again.");
                setMovieInfo(null);
            } finally {
                setIsFetchingDetails(false);
            }
        };

        fetchMovieDetails();
    }, [movieId]);

    const handleSubmit = async (formData: MovieFormData) => {
        setErrorMessage(null);
        setIsSubmitting(true);

        try {
            await movieService.updateMovie({ ...formData, id: Number(movieId) });
            window.scrollTo({ top: 0, behavior: "smooth" });
            setTriggerScroll(true);
            navigateWithSearchParams(`/${movieId}`);
        } catch (error) {
            console.error("Error updating a movie:", error);
            setErrorMessage("Failed to update a movie. Please try again.");
        } finally {
            setIsSubmitting(false);
        }
    };

    const handleClose = () => {
        if (movieInfo && movieId) {
            navigateWithSearchParams(`/${movieId}`);
        } else {
            navigateWithSearchParams(`/`);
        }
    };

    const showForm = !isFetchingDetails && !!movieInfo;
    const showErrorMessage = !isFetchingDetails && !!errorMessage;

    return (
        <Dialog title="Edit Movie" onClose={handleClose} className="max-w-[976px]">
            {isFetchingDetails && (
                <div className="flex justify-center items-center min-h-[700px] h-full">
                    <Loader />
                </div>
            )}

            {showErrorMessage && (
                <ErrorMessage message={errorMessage} onDismiss={() => setErrorMessage(null)} className="pb-6" />
            )}

            {showForm && (
                <MovieForm
                    initialMovieInfo={movieInfo}
                    onSubmit={handleSubmit}
                    isLoading={isSubmitting}
                    resetMode="restore"
                />
            )}
        </Dialog>
    );
};
