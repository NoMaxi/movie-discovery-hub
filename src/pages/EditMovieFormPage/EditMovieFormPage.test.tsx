import { ReactNode } from "react";
import { render, screen, waitFor, act } from "@testing-library/react";
import "@testing-library/jest-dom";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import { InitialMovieInfo, MovieFormData } from "@/types/common";
import { movieService } from "@/services/movieService";
import { mapAPIMovieDetailsToInitialMovieInfo } from "@/utils/movieMapping";
import { useNavigateWithSearchParams } from "@/hooks/useNavigateWithSearchParams/useNavigateWithSearchParams";
import { useScrollContext } from "@/contexts/ScrollContext/useScrollContext";
import { mockAPIMovie } from "@/mocks/movieData";
import { EditMovieFormPage } from "./EditMovieFormPage";

jest.mock("react-router-dom", () => ({
    ...jest.requireActual("react-router-dom"),
    useParams: jest.fn(),
}));
jest.mock("@/services/movieService");
jest.mock("@/hooks/useNavigateWithSearchParams/useNavigateWithSearchParams");
jest.mock("@/contexts/ScrollContext/useScrollContext");
jest.mock("@/utils/movieMapping");
jest.mock("@/components/common/Loader/Loader", () => ({
    Loader: () => <div data-testid="mock-loader">Loading...</div>,
}));

let mockDialogOnClose: () => void;
jest.mock("@/components/common/Dialog/Dialog", () => ({
    Dialog: ({
        title,
        children,
        onClose,
        className,
    }: {
        title: string;
        children: ReactNode;
        onClose: () => void;
        className?: string;
    }) => {
        mockDialogOnClose = onClose;
        return (
            <div data-testid="mock-dialog" className={className}>
                <h2 data-testid="mock-dialog-title">{title}</h2>
                <button data-testid="mock-dialog-close-button" onClick={onClose}>
                    Close
                </button>
                {children}
            </div>
        );
    },
}));

let mockMovieFormSubmit: (formData: MovieFormData) => void;
let lastMovieFormProps: { initialMovieInfo?: InitialMovieInfo | null; isLoading?: boolean; resetMode?: string } | null = null;

jest.mock("@/components/MovieForm/MovieForm", () => ({
    MovieForm: ({
        initialMovieInfo,
        onSubmit,
        isLoading,
        resetMode,
    }: {
        initialMovieInfo: InitialMovieInfo | null;
        onSubmit: (formData: MovieFormData) => void;
        isLoading?: boolean;
        resetMode?: string;
    }) => {
        mockMovieFormSubmit = onSubmit;
        lastMovieFormProps = { initialMovieInfo, isLoading, resetMode };
        return <form data-testid="mock-movie-form">Mock Movie Form</form>;
    },
}));

let mockErrorMessageDismiss: (() => void) | undefined;
jest.mock("@/components/common/ErrorMessage/ErrorMessage", () => ({
    ErrorMessage: ({
        message,
        onDismiss,
        className,
    }: {
        message: string | null;
        onDismiss: () => void;
        className?: string;
    }) => {
        mockErrorMessageDismiss = onDismiss;
        return message ? (
            <div data-testid="mock-error-message" className={className}>
                {message}
                <button data-testid="mock-error-dismiss-button" onClick={onDismiss}>
                    Dismiss
                </button>
            </div>
        ) : null;
    },
}));

const mockNavigateWithSearchParams = jest.fn();
const mockSetTriggerScroll = jest.fn();
const mockGetAPIMovieDetailsById = movieService.getAPIMovieDetailsById as jest.Mock;
const mockUpdateMovie = movieService.updateMovie as jest.Mock;
const mockMapAPIMovieDetailsToInitialMovieInfo = mapAPIMovieDetailsToInitialMovieInfo as jest.Mock;
const useParams = jest.requireMock("react-router-dom").useParams as jest.Mock;

const mockMovieId = "123";

const mockInitialMovieInfo: InitialMovieInfo = {
    title: "Test Movie",
    releaseDate: "2023-01-01",
    imageUrl: "path/to/poster.jpg",
    rating: 8.5,
    genres: ["Action", "Adventure"],
    duration: 120,
    description: "Test overview",
    id: 123,
};

const renderComponent = () => {
    const queryClient = new QueryClient();
    return render(
        <QueryClientProvider client={queryClient}>
            <EditMovieFormPage />
        </QueryClientProvider>,
    );
};

const delay = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms));

describe("EditMovieFormPage", () => {
    beforeEach(() => {
        jest.clearAllMocks();
        useParams.mockReturnValue({ movieId: mockMovieId });
        (useNavigateWithSearchParams as jest.Mock).mockReturnValue(mockNavigateWithSearchParams);
        (useScrollContext as jest.Mock).mockReturnValue({ setTriggerScroll: mockSetTriggerScroll });
        mockGetAPIMovieDetailsById.mockImplementation(async () => {
            await delay(10);
            return mockAPIMovie;
        });
        mockMapAPIMovieDetailsToInitialMovieInfo.mockReturnValue(mockInitialMovieInfo);
        window.scrollTo = jest.fn();
        mockErrorMessageDismiss = undefined;
        lastMovieFormProps = null;
    });

    test("Should show loader while fetching details", () => {
        renderComponent();
        expect(screen.getByTestId("mock-loader")).toBeInTheDocument();
        expect(screen.queryByTestId("mock-movie-form")).not.toBeInTheDocument();
        expect(screen.queryByTestId("mock-error-message")).not.toBeInTheDocument();
    });

    test("Should show error message if fetching details fails", async () => {
        const mockError = new Error("Fetch Error");
        mockGetAPIMovieDetailsById.mockRejectedValue(mockError);
        const consoleErrorSpy = jest.spyOn(console, "error").mockImplementation(() => {});

        renderComponent();

        await waitFor(() => {
            expect(screen.queryByTestId("mock-loader")).not.toBeInTheDocument();
        });
        expect(screen.getByTestId("mock-error-message")).toBeInTheDocument();
        expect(screen.getByTestId("mock-error-message")).toHaveTextContent(
            "Failed to load movie details. Please try again.",
        );
        expect(screen.queryByTestId("mock-movie-form")).not.toBeInTheDocument();
        expect(consoleErrorSpy).toHaveBeenCalledWith("Error fetching movie details during movie edit:", mockError);

        consoleErrorSpy.mockRestore();
    });

     test("Should show error message if movieId is missing", async () => {
        useParams.mockReturnValue({ movieId: undefined });
        renderComponent();

        await waitFor(() => {
            expect(screen.queryByTestId("mock-loader")).not.toBeInTheDocument();
        });
        expect(screen.getByTestId("mock-error-message")).toBeInTheDocument();
        expect(screen.getByTestId("mock-error-message")).toHaveTextContent("Movie ID not found.");
        expect(screen.queryByTestId("mock-movie-form")).not.toBeInTheDocument();
    });

    test("Should render correctly with Dialog and MovieForm after successful fetch", async () => {
        const { asFragment } = renderComponent();

        await waitFor(() => {
            expect(screen.queryByTestId("mock-loader")).not.toBeInTheDocument();
        });

        expect(screen.getByTestId("mock-dialog")).toBeInTheDocument();
        expect(screen.getByTestId("mock-dialog-title")).toHaveTextContent("Edit Movie");
        expect(screen.getByTestId("mock-movie-form")).toBeInTheDocument();
        expect(lastMovieFormProps?.initialMovieInfo).toEqual(mockInitialMovieInfo);
        expect(lastMovieFormProps?.resetMode).toBe("restore");
        expect(lastMovieFormProps?.isLoading).toBe(false);
        expect(screen.queryByTestId("mock-error-message")).not.toBeInTheDocument();
        expect(mockGetAPIMovieDetailsById).toHaveBeenCalledWith(Number(mockMovieId));
        expect(mockMapAPIMovieDetailsToInitialMovieInfo).toHaveBeenCalledWith(mockAPIMovie);
        expect(asFragment()).toMatchSnapshot();
    });

    test("Should handle successful form submission", async () => {
        const mockFormData: MovieFormData = {
            title: "Updated Movie",
            poster_path: "path/to/updated_image.jpg",
            genres: ["Drama"],
            overview: "An updated great movie",
            runtime: 130,
            release_date: "2024-01-01",
            vote_average: 9.0,
        };
        const expectedPayload = { ...mockFormData, id: Number(mockMovieId) };

        mockUpdateMovie.mockImplementation(async () => {
            await delay(10);
            return expectedPayload;
        });

        renderComponent();

        await waitFor(() => {
            expect(screen.getByTestId("mock-movie-form")).toBeInTheDocument();
        });

        await act(async () => {
            mockMovieFormSubmit(mockFormData);
        });

        await waitFor(() => expect(lastMovieFormProps?.isLoading).toBe(true));

        expect(mockUpdateMovie).toHaveBeenCalledTimes(1);
        expect(mockUpdateMovie).toHaveBeenCalledWith(expectedPayload);

        await waitFor(() => {
            expect(window.scrollTo).toHaveBeenCalledWith({ top: 0, behavior: "smooth" });
        });
        await waitFor(() => {
            expect(mockSetTriggerScroll).toHaveBeenCalledWith(true);
        });
        await waitFor(() => {
            expect(mockNavigateWithSearchParams).toHaveBeenCalledTimes(1);
        });
        expect(mockNavigateWithSearchParams).toHaveBeenCalledWith(`/${mockMovieId}`);

        expect(screen.queryByTestId("mock-error-message")).not.toBeInTheDocument();
    });

    test("Should handle failed form submission", async () => {
        const mockFormData: MovieFormData = {
            title: "Updated Movie",
            poster_path: "path/to/updated_image.jpg",
            genres: ["Drama"],
            overview: "An updated great movie",
            runtime: 130,
            release_date: "2024-01-01",
            vote_average: 9.0,
        };
        const mockError = new Error("Update API Error");

        mockUpdateMovie.mockImplementation(async () => {
            await delay(10);
            throw mockError;
        });
        const consoleErrorSpy = jest.spyOn(console, "error").mockImplementation(() => {});

        renderComponent();

        await waitFor(() => {
            expect(screen.getByTestId("mock-movie-form")).toBeInTheDocument();
        });

        await act(async () => {
            mockMovieFormSubmit(mockFormData);
        });

        await waitFor(() => expect(lastMovieFormProps?.isLoading).toBe(true));

        expect(mockUpdateMovie).toHaveBeenCalledTimes(1);
        expect(mockUpdateMovie).toHaveBeenCalledWith({ ...mockFormData, id: Number(mockMovieId) });

        await waitFor(() => {
            expect(consoleErrorSpy).toHaveBeenCalledWith("Error updating a movie:", mockError);
        });
        await waitFor(() => {
            expect(screen.getByTestId("mock-error-message")).toBeInTheDocument();
        });
        expect(screen.getByTestId("mock-error-message")).toHaveTextContent(
            "Failed to update a movie. Please try again.",
        );

        expect(mockNavigateWithSearchParams).not.toHaveBeenCalled();
        expect(window.scrollTo).not.toHaveBeenCalled();
        expect(mockSetTriggerScroll).not.toHaveBeenCalled();

        consoleErrorSpy.mockRestore();
    });

    test("Should handle dialog close when movie info is loaded", async () => {
        renderComponent();

        await waitFor(() => {
            expect(screen.getByTestId("mock-movie-form")).toBeInTheDocument();
        });

        await act(async () => {
            mockDialogOnClose();
        });

        await waitFor(() => {
            expect(mockNavigateWithSearchParams).toHaveBeenCalledTimes(1);
        });
        expect(mockNavigateWithSearchParams).toHaveBeenCalledWith(`/${mockMovieId}`);
    });

     test("Should handle dialog close when movie info failed to load", async () => {
        mockGetAPIMovieDetailsById.mockRejectedValue(new Error("Fetch Error"));
        renderComponent();

        await waitFor(() => {
            expect(screen.getByTestId("mock-error-message")).toBeInTheDocument();
        });

        await act(async () => {
            mockDialogOnClose();
        });

        await waitFor(() => {
            expect(mockNavigateWithSearchParams).toHaveBeenCalledTimes(1);
        });
        expect(mockNavigateWithSearchParams).toHaveBeenCalledWith("/");
    });

     test("Should handle dialog close when movieId is missing", async () => {
        useParams.mockReturnValue({ movieId: undefined });
        renderComponent();

        await waitFor(() => {
            expect(screen.getByTestId("mock-error-message")).toBeInTheDocument();
        });

        await act(async () => {
            mockDialogOnClose();
        });

        await waitFor(() => {
            expect(mockNavigateWithSearchParams).toHaveBeenCalledTimes(1);
        });
        expect(mockNavigateWithSearchParams).toHaveBeenCalledWith("/");
    });

    test("Should handle fetch error message dismissal", async () => {
        mockGetAPIMovieDetailsById.mockRejectedValue(new Error("Fetch Error"));
        jest.spyOn(console, "error").mockImplementation(() => {});
        renderComponent();

        await waitFor(() => {
            expect(screen.getByTestId("mock-error-message")).toBeInTheDocument();
        });

        expect(mockErrorMessageDismiss).toBeDefined();

        await act(async () => {
            if (mockErrorMessageDismiss) {
                mockErrorMessageDismiss();
            }
        });

        await waitFor(() => {
            expect(screen.queryByTestId("mock-error-message")).not.toBeInTheDocument();
        });
        jest.restoreAllMocks();
    });

    test("Should handle submit error message dismissal", async () => {
        const mockError = new Error("Update API Error");
        mockUpdateMovie.mockRejectedValue(mockError);
        jest.spyOn(console, "error").mockImplementation(() => {});

        renderComponent();

        await waitFor(() => {
            expect(screen.getByTestId("mock-movie-form")).toBeInTheDocument();
        });

        await act(async () => {
            mockMovieFormSubmit({} as MovieFormData);
        });

        await waitFor(() => {
            expect(screen.getByTestId("mock-error-message")).toBeInTheDocument();
        });

        expect(mockErrorMessageDismiss).toBeDefined();

        await act(async () => {
            if (mockErrorMessageDismiss) {
                mockErrorMessageDismiss();
            }
        });

        await waitFor(() => {
            expect(screen.queryByTestId("mock-error-message")).not.toBeInTheDocument();
        });
        jest.restoreAllMocks();
    });
});
