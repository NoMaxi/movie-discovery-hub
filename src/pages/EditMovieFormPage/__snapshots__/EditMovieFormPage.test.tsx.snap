// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditMovieFormPage Should render correctly with Dialog and MovieForm after successful fetch 1`] = `
<DocumentFragment>
  <div
    class="max-w-[976px]"
    data-testid="mock-dialog"
  >
    <h2
      data-testid="mock-dialog-title"
    >
      Edit Movie
    </h2>
    <button
      data-testid="mock-dialog-close-button"
    >
      Close
    </button>
    <form
      data-testid="mock-movie-form"
    >
      Mock Movie Form
    </form>
  </div>
</DocumentFragment>
`;
